triqs/operators/many_body_operator.hpp

```diff
@@ -285,28 +285,32 @@ namespace operators {
    return res;
   }
 
-  // real
-  friend many_body_operator_generic real(many_body_operator_generic const& op) {
+  // Transform operator by applying a given function to each monomial
+  template <typename Lambda>
+  friend many_body_operator_generic foreach_monomial(many_body_operator_generic const& op, Lambda L) {
    many_body_operator_generic res;
-   using triqs::utility::real;
    using triqs::utility::is_zero;
    for (auto const& x : op) {
-    auto c = real(x.coef);
+    auto c = L(x.monomial, x.coef);
     if(!is_zero(c)) res.monomials.insert({x.monomial, c});
    }
    return res;
   }
 
+  // real
+  friend many_body_operator_generic real(many_body_operator_generic const& op) {
+   return foreach_monomial(op, [](monomial_t const&, scalar_t c) {
+    using triqs::utility::real;
+    return real(c);
```

Comment:
Compilation of your variant will fail if `triqs::utility::real(scalar_t)` is not defined.
There is still a chance, however, that `real(scalar_t)` is found by argument-dependent lookup, which makes compilation of `real(c)` possible.